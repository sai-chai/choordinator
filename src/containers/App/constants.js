export const RANGE_TUPLET = ['C0', 'C8'];

export const DEFAULT_TONIC = 'C4';
