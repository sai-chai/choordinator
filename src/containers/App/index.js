import React, { useState, useMemo } from 'react';
import styled from 'styled-components';

import { entries as scaleDictionary } from '@tonaljs/scale-dictionary';
import { chromatic } from '@tonaljs/range';

import PianoRoll from 'containers/PianoRoll';
import Select from 'components/Select';
import ArrowSelect from 'components/ArrowSelect';

import { RANGE_TUPLET, DEFAULT_TONIC } from './constants';


function App (props) {
   const [scale, setScale] = useState("major");
   const [tonic, setTonic] = useState(DEFAULT_TONIC);

   // For scale dropdown, e.g. major, minor, mixolydian, etc.
   const scaleNames = useMemo(
      () => scaleDictionary().reduce(
         (acc, curr) => {
            acc.push(curr.name);
            return acc;
         }, []
      ).sort()
   );

   // For tonic selector, e.g. C5, Ab4, G1, etc.
   const tonicList = useMemo(() => chromatic(RANGE_TUPLET));

   return (
      <Container>
         <Header>
            <Title>
               The Choordinator
            </Title>
            <Subtitle>
               Playing the white keys never sounded so good...
            </Subtitle>
            {/* <InfoButton onClick={onInfoClick} /> */}
         </Header>
         <Main>
            <Settings>
               <div>
                  <Label>Tonic note:</Label>
                  <Label>Scale:</Label>
               </div>
               <div>
                  <ArrowSelect
                     name="tonic"
                     value={tonic}
                     options={tonicList}
                     setValue={setTonic}
                  />
                  <Select
                     name="scale"
                     value={scale}
                     options={scaleNames}
                     setValue={setScale}
                  />
               </div>
            </Settings>
            <PianoRoll tonic={tonic} scale={scale} />
         </Main>
      </Container>
   );
}

const Container = styled.div`
   background-color: #F3DEC1;
   min-height: 100vh;
   font-family: 'Helvetica', Arial, sans-serif;
`;

const Header = styled.header.attrs({
   role: 'navigation',
})`
   color: #fff;
   background-color: #77474B;
   height: 50px;
   padding: 0 50px;
   display: flex;
   flex-direction: row;
   justify-content: space-between;
`;

const Title = styled.h1`
   font-size: 2.4rem;
   line-height: 5.0rem;
   font-style: italic;
   font-weight: 500;
`;

const Subtitle = styled.h3`
   font-size: 1.6rem;
   line-height: 5.0rem;
   font-family: 'Courier New', Courier, monospace;
   font-style: italic;
   font-weight: 700;
   @media (max-width: 770px) {
      display: none;
   }
`;

const Main = styled.main`
   max-width: 1000px;
   margin: 0 auto;
   padding: 0 0 36px;
`;

const Settings = styled.form`
   display: flex;
   flex-direction: row;
   padding: 2.4rem;
   justify-content: center;
   > div {
      display: flex;
      flex-direction: column;
      margin: 0 0.8rem;
      &:first-child {
         margin-left: 0;
      }
      &:last-child {
         margin-right: 0;
      }
   }
`;

const Label = styled.label`
   line-height: 3rem;
   font-size: 1.6rem;
   margin: 0.8rem 0;
   &:first-child {
      margin-top: 0;
   }
   &:last-child {
      margin-bottom: 0;
   }
`;

export default App;
