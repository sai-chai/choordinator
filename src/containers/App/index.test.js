jest.mock('soundfont-player');
import { AudioContext } from 'web-audio-mock-api';
import flushPromises from 'flush-promises';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { shallow } from 'enzyme';
import App from './index';

describe('<PianoRoll />', () => {
   window.AudioContext = AudioContext;

   it('smoke test', async () => {
      const wrapper = shallow(<App />);
      await act(async () => {
         await flushPromises();
      });
      expect(wrapper).toMatchSnapshot();
   });
});
