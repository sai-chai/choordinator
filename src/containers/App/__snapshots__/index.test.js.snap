// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PianoRoll /> smoke test 1`] = `
<App__Container>
  <App__Header>
    <App__Title>
      The Choordinator
    </App__Title>
    <App__Subtitle>
      Playing the white keys never sounded so good...
    </App__Subtitle>
  </App__Header>
  <App__Main>
    <App__Settings>
      <div>
        <App__Label>
          Tonic note:
        </App__Label>
        <App__Label>
          Scale:
        </App__Label>
      </div>
      <div>
        <ArrowSelect
          name="tonic"
          options={
            Array [
              "C0",
              "Db0",
              "D0",
              "Eb0",
              "E0",
              "F0",
              "Gb0",
              "G0",
              "Ab0",
              "A0",
              "Bb0",
              "B0",
              "C1",
              "Db1",
              "D1",
              "Eb1",
              "E1",
              "F1",
              "Gb1",
              "G1",
              "Ab1",
              "A1",
              "Bb1",
              "B1",
              "C2",
              "Db2",
              "D2",
              "Eb2",
              "E2",
              "F2",
              "Gb2",
              "G2",
              "Ab2",
              "A2",
              "Bb2",
              "B2",
              "C3",
              "Db3",
              "D3",
              "Eb3",
              "E3",
              "F3",
              "Gb3",
              "G3",
              "Ab3",
              "A3",
              "Bb3",
              "B3",
              "C4",
              "Db4",
              "D4",
              "Eb4",
              "E4",
              "F4",
              "Gb4",
              "G4",
              "Ab4",
              "A4",
              "Bb4",
              "B4",
              "C5",
              "Db5",
              "D5",
              "Eb5",
              "E5",
              "F5",
              "Gb5",
              "G5",
              "Ab5",
              "A5",
              "Bb5",
              "B5",
              "C6",
              "Db6",
              "D6",
              "Eb6",
              "E6",
              "F6",
              "Gb6",
              "G6",
              "Ab6",
              "A6",
              "Bb6",
              "B6",
              "C7",
              "Db7",
              "D7",
              "Eb7",
              "E7",
              "F7",
              "Gb7",
              "G7",
              "Ab7",
              "A7",
              "Bb7",
              "B7",
              "C8",
            ]
          }
          setValue={[Function]}
          value="C4"
        />
        <Select
          name="scale"
          options={
            Array [
              "\\"aeolian b5",
              "aeolian",
              "altered",
              "arabian",
              "augmented",
              "augmented heptatonic",
              "balinese",
              "bebop",
              "bebop locrian",
              "bebop major",
              "bebop minor",
              "blues",
              "chinese",
              "chromatic",
              "composite blues",
              "diminished",
              "diminished whole tone",
              "dominant",
              "dominant diminished",
              "dorian",
              "dorian #4",
              "dorian b2",
              "double harmonic lydian",
              "double harmonic major",
              "egyptian",
              "enigmatic",
              "flamenco",
              "flat six pentatonic",
              "flat three pentatonic",
              "gypsy",
              "half-diminished",
              "half-whole diminished",
              "harmonic major",
              "harmonic minor",
              "hindu",
              "hirajoshi",
              "hungarian major",
              "hungarian minor",
              "ichikosucho",
              "in-sen",
              "indian",
              "ionian",
              "ionian #5",
              "ionian augmented",
              "ionian pentatonic",
              "iwato",
              "kafi raga",
              "kumoi",
              "kumoijoshi",
              "leading whole tone",
              "locrian",
              "locrian #2",
              "locrian 6",
              "locrian major",
              "locrian natural 6",
              "locrian pentatonic",
              "locrian sharp 6",
              "lydian",
              "lydian #5P pentatonic",
              "lydian #9",
              "lydian augmented",
              "lydian b7",
              "lydian diminished",
              "lydian dominant",
              "lydian dominant pentatonic",
              "lydian minor",
              "lydian pentatonic",
              "major",
              "major #5",
              "major augmented",
              "major blues",
              "major pentatonic",
              "malkos raga",
              "melodic minor",
              "melodic minor fifth mode",
              "melodic minor second mode",
              "minor",
              "minor #7M pentatonic",
              "minor bebop",
              "minor blues",
              "minor hexatonic",
              "minor pentatonic",
              "minor seven flat five pentatonic",
              "minor six diminished",
              "minor six pentatonic",
              "mixolydian",
              "mixolydian b6",
              "mixolydian pentatonic",
              "mystery #1",
              "neopolitan major",
              "neopolitan major pentatonic",
              "oriental",
              "overtone",
              "pelog",
              "pentatonic",
              "persian",
              "phrygian",
              "phrygian #6",
              "phrygian dominant",
              "phrygian major",
              "piongio",
              "pomeroy",
              "prometheus",
              "prometheus neopolitan",
              "purvi raga",
              "ritusen",
              "romanian minor",
              "scriabin",
              "six tone symmetric",
              "spanish",
              "spanish heptatonic",
              "super locrian",
              "super locrian pentatonic",
              "superlocrian bb7",
              "todi raga",
              "ultralocrian",
              "vietnamese 1",
              "vietnamese 2",
              "whole tone",
              "whole tone pentatonic",
              "whole-half diminished",
              "Â·superlocrian diminished",
            ]
          }
          setValue={[Function]}
          value="major"
        />
      </div>
    </App__Settings>
    <PianoRoll
      scale="major"
      tonic="C4"
    />
  </App__Main>
</App__Container>
`;
